### PR TITLE
[contrib] Refactor `incidence_analysis` to cache a graph instead of a matrix

### PR DESCRIPTION
**Pull request by Robbybp**
## EDIT: Dependence of #2715, #2716, #2723, and #2727
- #2716 incorporates #2715
- #2727 incorporates #2716 and #2723
- #2723 is small and NFC
- #2727 changes behavior

In summary, merging #2727 will incorporate all the others, but they are broken up to be easier to review.

## Summary/Motivation:
`IncidenceGraphInterface` accepts a model and caches the incidence matrix (in COO format) of variables and constraints. This allows calling `dulmage_mendelsohn`, etc. without re-constructing the matrix, but does not allow fast look-ups of "adjacent" constraints and variables, i.e. getting all constraints that contain a variable or all variables that participate in a constraint. This would require storing the incidence matrix in *both* CSR and CSC formats, or alternatively storing the bipartite incidence graph as an adjacency list. This PR refactors `incidence_analysis` to cache the bipartite incidence graph as a NetworkX Graph instead of caching a SciPy COO matrix, allowing a `get_adjacent_to` method that quickly looks up adjacent variables/constraints.

## Changes proposed in this PR:
- `incidence_graph` attribute on `IncidenceGraphInterface`
- `incidence_matrix` attribute is now a "property" that computes the matrix from the cached graph. This means that this attribute still works, but is now an expensive operation.
- Refactor "algorithm methods", e.g. `dulmage_mendelsohn`, and the underlying implementations so we do not convert the cached graph to a COO matrix, just to convert it back to a graph
- Add `get_adjacent_to` method

This PR also runs Black on `incidence_analysis`, for good measure.

## Next steps:
This PR sets the stage for several improvements I'd like to make to `incidence_analysis`, which will be subsequent PRs:
1. Add a plotting method such as https://github.com/Pyomo/pyomo/pull/2251 to `IncidenceGraphInterface`
2. Update the API to remove public attributes from `IncidenceGraphInterface`. I would also like to update the `block_triangularize` and `get_diagonal_blocks` method names and APIs, as I think the maps returned by `block_triangularize` are confusing, and the name `get_diagonal_blocks` is an incomplete misnomer.
3. Add documentation

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
4. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

**Comment by Robbybp** [comment]
win/3.9 teest failure appears unrelated

**Comment by mrmundt** [comment]
> win/3.9 teest failure appears unrelated

It is unrelated. We're looking into this failure - it's been consistent across all new jobs run today.

**Comment by Robbybp** [comment]
@mrmundt Thanks!

File: pyomo/contrib/incidence_analysis/interface.py [hunk 1]
```diff
@@ -72,21 +86,78 @@ def get_incidence_graph(variables, constraints, include_fixed=True):
...
     return graph
 
 
+def extract_bipartite_subgraph(graph, nodes0, nodes1):
+    """Return the bipartite subgraph of a graph.
+
+    Two lists of nodes to project onto must be provided. These will correspond
+    to the "bipartite sets" in the subgraph. If the two sets provided have
+    M and N nodes, the subgraph will have nodes 0 through M+N, with the first
+    M corresponding to the first set provided and the last M corresponding
```
Comment (mrmundt) [inline pyomo/contrib/incidence_analysis/interface.py:111]: Maybe typo? `First M ... last N`

File: pyomo/contrib/incidence_analysis/interface.py [hunk 2]
```diff
@@ -72,21 +86,78 @@ def get_incidence_graph(variables, constraints, include_fixed=True):
...
+    to the second set.
+
+    Parameters
+    ----------
+    graph: NetworkX Graph
+        The graph from which a subgraph is extracted
+    nodes0: list
+        A list of nodes in the original graph that will form the first
+        bipartite set of the projected graph (and have ``bipartite=0``
```
Comment (mrmundt) [inline pyomo/contrib/incidence_analysis/interface.py:120]: Small typo: missing close parenthesis.

File: pyomo/contrib/incidence_analysis/interface.py [hunk 3]
```diff
@@ -72,21 +86,78 @@ def get_incidence_graph(variables, constraints, include_fixed=True):
...
+    nodes1: list
+        A list of nodes in the original graph that will form the second
+        bipartite set of the projected graph (and have ``bipartite=1``
```
Comment (mrmundt) [inline pyomo/contrib/incidence_analysis/interface.py:123]: Small typo: missing close parenthesis.

File: pyomo/contrib/incidence_analysis/interface.py [hunk 4]
```diff
@@ -208,56 +288,59 @@ def __init__(
...
-            self.incidence_matrix = get_structural_incidence_matrix(
-                    self.variables,
-                    self.constraints,
-                    )
+                (con, i) for i, con in enumerate(self.constraints)
+            )
+            self.incidence_graph = get_bipartite_incidence_graph(
+                self.variables,
+                self.constraints,
+                # TODO: include_fixed=include_fixed?
```
Comment (mrmundt) [inline pyomo/contrib/incidence_analysis/interface.py:299]: Just checking: a real TODO or residual note to yourself?

File: pyomo/contrib/incidence_analysis/interface.py [hunk 5]
```diff
@@ -268,23 +351,107 @@ def _extract_submatrix(self, variables, constraints):
...
-                    shape=(M, N),
-                    )
+                (new_data, (new_row, new_col)),
+                shape=(M, N),
+            )
+
+    def _extract_subgraph(self, variables, constraints):
+        if self.incidence_graph is None:
+            return get_bipartite_incidence_graph(
+                # Does include_fixed matter here if I'm providing the variables?
```
Comment (mrmundt) [inline pyomo/contrib/incidence_analysis/interface.py:361]: Comment for yourself?

**Review by mrmundt — CHANGES_REQUESTED** [review]
There are some small typos and questions that I have, but otherwise, this is looking great.

File: pyomo/contrib/incidence_analysis/interface.py (hunk 4 above)
Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py:299]: Residual note. However, `include_fixed` is not necessary here, as we have already filtered above in the `_generate_variables_in_constraints` call.

File: pyomo/contrib/incidence_analysis/interface.py (hunk 1 above)
Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py]: Thanks, fixed!

File: pyomo/contrib/incidence_analysis/interface.py (hunk 5 above)
Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py]: Comment for myself. Note that this does change behavior. However, I feel that if variables are explicitly provided, they should be part of the graph regardless of whether they are fixed.

Comment (Robbybp) [reply pyomo/contrib/incidence_analysis/interface.py]: To clarify: removing the `include_fixed=True` argument changes behavior (to be more rational, in my opinion).

**Review by mrmundt — APPROVED** [review]

**Review by blnicho — APPROVED** [review]

**Pull request merged**